### PR TITLE
feat(eslint-plugin): [no-empty-func] private/protected construct

### LINKED ISSUE #1178: [no-empty-function] private constructors should not error
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-empty-function": "on",
  }
}
```

```TS
class Foo {
  private constructor() {}
}
```

**Expected Result**

No warning because a private no-args constructor is useful to force instantiation of a class from static methods.

**Actual Result**

The rule warns that the private constructor is useless.
```
error  Unexpected empty constructor  @typescript-eslint/no-empty-function
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.6.1` |
| `@typescript-eslint/parser`        | `2.6.1` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `6.6.0` |
| `node`                             | `12.10` |
| `npm`                              | `6.11.3` |

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #1178

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/no-empty-function.md [hunk 1]
```diff
@@ -44,4 +44,36 @@ See the [ESLint documentation](https://eslint.org/docs/rules/no-empty-function)
 }
 ```
 
+## Options
+
+This rule has an object option:
+
+- `allow` (`string[]`)
+  - `"protected-constructors"` - Protected class constructors.
+  - `"private-constructors"` - Private class constructors.
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-empty-function.md]: nit, would probably be good to point the user toward the "extra" options
```suggestion
  - `"private-constructors"` - Private class constructors.
  - [See the other options allowed](https://github.com/eslint/eslint/blob/master/docs/rules/no-empty-function.md#options)
```

File: packages/eslint-plugin/src/rules/no-empty-function.ts [hunk 2]
```diff
@@ -74,30 +89,45 @@ export default util.createRule<Options, MessageIds>({
...
     ): boolean {
-      // Check TypeScript specific nodes
-      return (
-        isConstructor(node) && isBodyEmpty(node) && hasParameterProperties(node)
-      );
+      const parent = node.parent;
+      if (
+        isBodyEmpty(node) &&
+        parent &&
+        parent.type === 'MethodDefinition' &&
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-empty-function.ts]: nit: optional chaining.
TS will refine the type of parent to be non-nullish for the rest of the scope.

```
parent?.type === 'MethodDefinition' &&
```

File: packages/eslint-plugin/docs/rules/no-empty-function.md (hunk 1 above)
Comment (a-tarasyuk) [reply packages/eslint-plugin/docs/rules/no-empty-function.md]: Yes, you are right. Thanks.

File: packages/eslint-plugin/src/rules/no-empty-function.ts (hunk 2 above)
Comment (a-tarasyuk) [reply packages/eslint-plugin/src/rules/no-empty-function.ts]: Good point. Thanks.

**Pull request merged**